File: api/yahoo_spot.py
"""
Yahoo Finance spot-quote proxy, run as a Vercel Python serverless function.

We call this from the Node.js cron when Stooq lacks the user's actual
exchange listing (today: just IEMM held on Euronext Amsterdam in EUR;
Stooq only carries the LSE GBP twin under EIMI). Direct fetches against
Yahoo's chart endpoint from Vercel's IPs get 429'd within seconds; the
`yfinance` library handles the cookie + crumb session Yahoo expects from
real clients, which lifts the rate limit enough for our daily-cron
volume.

Auth: Bearer ${CRON_SECRET}. Query: ?symbol=<internal>[,<internal>...].
Returns: { quotes: [{symbol,date,close,currency}], errors: [{symbol,error}] }.
"""

import json
import os
from http.server import BaseHTTPRequestHandler
from urllib.parse import parse_qs, urlparse


# Internal canonical ticker → Yahoo exchange-suffixed symbol. Mirrors
# src/lib/quotes/symbol-map.ts for the entries we actually route here.
SYMBOL_MAP = {
    "IEMM": "IEMM.AS",  # iShares MSCI EM UCITS — Amsterdam EUR class
}


class handler(BaseHTTPRequestHandler):
    def do_GET(self) -> None:
        if not self._authed():
            self._json(401, {"error": "unauthorized"})
            return

        symbols = self._parse_symbols()
        if not symbols:
            self._json(400, {"error": "missing symbol query param"})
            return

        try:
            import yfinance as yf
        except Exception as e:  # noqa: BLE001
            self._json(500, {"error": f"yfinance_import_failed: {e}"})
            return

        quotes = []
        errors = []
        for internal in symbols:
            yahoo_symbol = SYMBOL_MAP.get(internal, internal)
            try:
                ticker = yf.Ticker(yahoo_symbol)
                hist = ticker.history(period="5d", auto_adjust=False)
                if hist.empty:
                    errors.append({"symbol": internal, "error": "no_data"})
                    continue
                last_row = hist.iloc[-1]
                last_date = hist.index[-1].strftime("%Y-%m-%d")
                try:
                    currency = ticker.fast_info["currency"] or "EUR"
                except Exception:  # noqa: BLE001
                    currency = "EUR"
                quotes.append(
                    {
                        "symbol": internal,
                        "date": last_date,
                        "close": f"{float(last_row['Close']):.6f}",
                        "currency": currency,
                    }
                )
            except Exception as e:  # noqa: BLE001
                errors.append(
                    {"symbol": internal, "error": f"{type(e).__name__}: {e}"}
                )

        self._json(200, {"quotes": quotes, "errors": errors})

    def _authed(self) -> bool:
        auth = self.headers.get("Authorization", "")
        secret = os.environ.get("CRON_SECRET", "")
        return bool(secret) and auth == f"Bearer {secret}"

    def _parse_symbols(self) -> list[str]:
        qs = parse_qs(urlparse(self.path).query)
        raw = qs.get("symbol", [""])[0]
        return [s.strip() for s in raw.split(",") if s.strip()]

    def _json(self, status: int, body: dict) -> None:
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.end_headers()
        self.wfile.write(json.dumps(body).encode())

File: src/lib/quotes/yahoo-spot.ts
import { fetchYahooHistory } from "./history";

/**
 * Latest spot quote from Yahoo using the v8/chart endpoint (the v7/quote
 * endpoint regularly 429s anonymous callers from Vercel IPs; the chart
 * endpoint is far more tolerant). We ask for a 5-day window and take the
 * most recent non-null close — same currency the chart endpoint reports.
 */
export type YahooSpot = { symbol: string; date: string; close: string; currency: string };

export type YahooSpotResult =
  | { ok: true; quote: YahooSpot }
  | { ok: false; symbol: string; error: string };

export async function fetchYahooSpot(symbol: string): Promise<YahooSpotResult> {
  try {
    const rows = await fetchYahooHistory(symbol, "5d");
    if (!rows.length) return { ok: false, symbol, error: "no rows" };
    // History is chronological; the last entry is the most recent close.
    const last = rows[rows.length - 1];
    return { ok: true, quote: { symbol, date: last.date, close: last.close, currency: last.currency } };
  } catch (e) {
    return { ok: false, symbol, error: (e as Error).message };
  }
}

export async function fetchYahooSpots(symbols: string[]): Promise<{ quotes: YahooSpot[]; errors: Array<{ symbol: string; error: string }> }> {
  const quotes: YahooSpot[] = [];
  const errors: Array<{ symbol: string; error: string }> = [];
  for (let i = 0; i < symbols.length; i++) {
    const r = await fetchYahooSpot(symbols[i]);
    if (r.ok) quotes.push(r.quote);
    else errors.push({ symbol: r.symbol, error: r.error });
    // Throttle to stay friendly with Yahoo's anonymous rate limit; cron
    // budget is 60 s and we only ever call this for the handful of
    // symbols Stooq can't cover (today: just IEMM).
    if (i < symbols.length - 1) {
      await new Promise((r) => setTimeout(r, 700));
    }
  }
  return { quotes, errors };
}

/**
 * Symbols where Yahoo is the authoritative source instead of Stooq —
 * typically because Stooq lacks the specific exchange listing the user
 * actually holds, and we'd otherwise approximate via a sibling listing in
 * a different currency.
 */
export const YAHOO_PRIMARY_SYMBOLS = new Set<string>([
  "IEMM", // iShares MSCI EM UCITS — Amsterdam EUR class (IEMM.AS). Stooq
          // only has the LSE GBP twin under EIMI, which tracks the same NAV
          // but in a different currency and slightly different price.
]);
